processing/src/main/java/io/druid/query/PrioritizedExecutorService.java

```diff
@@ -185,89 +197,130 @@ public int getQueueSize()
+  {
+    return new PrioritizedListenableFutureTask<>(task, priority, position);
+  }
 
-    PrioritizedListenableFutureTask(ListenableFutureTask<V> delegate, int priority)
-    {
-      this.delegate = delegate;
-      this.priority = priority;
-    }
+  private final ListenableFutureTask<V> delegate;
+  private final int priority;
+  private final long insertionPlace;
+  private final Comparator<PrioritizedListenableFutureTask> comparator;
 
-    @Override
-    public void run()
-    {
-      delegate.run();
-    }
+  PrioritizedListenableFutureTask(ListenableFutureTask<V> delegate, int priority, long position)
+  {
+    this.delegate = delegate;
+    this.priority = priority;
+    this.insertionPlace = position; // Long.MAX_VALUE will always be "highest"
+    this.comparator = PRIORITY_COMPARATOR;
+  }
 
-    @Override
-    public boolean cancel(boolean mayInterruptIfRunning)
-    {
-      return delegate.cancel(mayInterruptIfRunning);
-    }
+  @Override
+  public void run()
+  {
+    delegate.run();
+  }
 
-    @Override
-    public boolean isCancelled()
-    {
-      return delegate.isCancelled();
-    }
+  @Override
+  public boolean cancel(boolean mayInterruptIfRunning)
+  {
+    return delegate.cancel(mayInterruptIfRunning);
+  }
 
-    @Override
-    public boolean isDone()
-    {
-      return delegate.isDone();
-    }
+  @Override
+  public boolean isCancelled()
+  {
+    return delegate.isCancelled();
+  }
 
-    @Override
-    public V get() throws InterruptedException, ExecutionException
-    {
-      return delegate.get();
-    }
+  @Override
+  public boolean isDone()
+  {
+    return delegate.isDone();
+  }
 
-    @Override
-    public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException
-    {
-      return delegate.get(timeout, unit);
-    }
+  @Override
+  public V get() throws InterruptedException, ExecutionException
+  {
+    return delegate.get();
+  }
 
-    @Override
-    public void addListener(Runnable listener, Executor executor)
-    {
-      delegate.addListener(listener, executor);
-    }
+  @Override
+  public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException
+  {
+    return delegate.get(timeout, unit);
+  }
 
-    @Override
-    public int getPriority()
-    {
-      return priority;
-    }
+  @Override
+  public void addListener(Runnable listener, Executor executor)
+  {
+    delegate.addListener(listener, executor);
+  }
 
-    @Override
-    public int compareTo(PrioritizedListenableFutureTask otherTask)
-    {
-      return Ints.compare(otherTask.getPriority(), getPriority());
-    }
+  @Override
+  public int getPriority()
+  {
+    return priority;
+  }
+
+  protected long getInsertionPlace()
+  {
+    return insertionPlace;
+  }
+
+  @Override
+  public int compareTo(PrioritizedListenableFutureTask otherTask)
+  {
+    return comparator.compare(this, otherTask);
```

Comment:
the "comparator" field seems to not be doing much, this could refer to PRIORITY_COMPARATOR